pkg/channel/consolidated/kafka/admin.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kafka
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/Shopify/sarama"
+	"knative.dev/pkg/logging"
+)
+
+var mutex sync.Mutex
+
+type ClusterAdminFactory func() (sarama.ClusterAdmin, error)
+
+type AdminClient interface {
+	// ListConsumerGroups Lists the consumer groups
+	ListConsumerGroups() ([]string, error)
+}
+
+// AdminClientManager manages a ClusterAdmin connection and recreates one when needed
+// it is made to overcome https://github.com/Shopify/sarama/issues/1162
+type AdminClientManager struct {
+	logger    *zap.SugaredLogger
+	caFactory ClusterAdminFactory
+	ca        sarama.ClusterAdmin
+}
+
+func NewAdminClient(ctx context.Context, caFactory ClusterAdminFactory) (AdminClient, error) {
+	logger := logging.FromContext(ctx)
+	logger.Info("Creating a new AdminClient")
+	kafkaClusterAdmin, err := caFactory()
+	if err != nil {
+		logger.Error("error while creating ClusterAdmin", err)
+		return &AdminClientManager{}, err
+	}
+	return &AdminClientManager{
+		logger:    logger,
+		caFactory: caFactory,
+		ca:        kafkaClusterAdmin,
+	}, nil
+}
+
+func (c *AdminClientManager) ListConsumerGroups() ([]string, error) {
+	c.logger.Info("Attempting to list consumer group")
+	mutex.Lock()
+	defer mutex.Unlock()
+	r := 0
+	// This gives us around ~13min of exponential backoff
+	max := 13
+	cgsMap, err := c.ca.ListConsumerGroups()
+	for err != nil && r <= max {
+		// presuming we can reconnect
+		t := int(math.Pow(2, float64(r)) * 100)
+		d := time.Duration(t) * time.Millisecond
+		c.logger.Error("listing consumer group failed. Refreshing the ClusterAdmin and retrying.",
+			zap.Error(err),
+			zap.Duration("retry after", d),
+			zap.Int("Retry attempt", r),
+			zap.Int("Max retries", max),
+		)
+		time.Sleep(d)
+		c.ca, err = c.caFactory()
+		r += 1
+		if err != nil {
+			// skip this attempt
+			continue
+		}
+		cgsMap, err = c.ca.ListConsumerGroups()
```

Comment:
Why locking? Isn't the ca already thread safe?

Reply:
I was presuming there would be a race between lines 71 and 83. I did a hacky test on my machine and indeed there is

```
WARNING: DATA RACE
Write at 0x00c00022e070 by goroutine 13:
  knative.dev/eventing-kafka/pkg/channel/consolidated/kafka.(*AdminClientManager).ListConsumerGroups()
      /home/abd4lla/Workspace/redhat/src/knative.dev/eventing-kafka/pkg/channel/consolidated/kafka/admin.go:83 +0x6bd
  knative.dev/eventing-kafka/pkg/channel/consolidated/kafka.doList()
      /home/abd4lla/Workspace/redhat/src/knative.dev/eventing-kafka/pkg/channel/consolidated/kafka/admin_test.go:86 +0x48
  knative.dev/eventing-kafka/pkg/channel/consolidated/kafka.TestAdminClient.func2()
      /home/abd4lla/Workspace/redhat/src/knative.dev/eventing-kafka/pkg/channel/consolidated/kafka/admin_test.go:63 +0x50

Previous read at 0x00c00022e070 by goroutine 16:
  knative.dev/eventing-kafka/pkg/channel/consolidated/kafka.(*AdminClientManager).ListConsumerGroups()
      /home/abd4lla/Workspace/redhat/src/knative.dev/eventing-kafka/pkg/channel/consolidated/kafka/admin.go:71 +0xe8
  knative.dev/eventing-kafka/pkg/channel/consolidated/kafka.doList()
      /home/abd4lla/Workspace/redhat/src/knative.dev/eventing-kafka/pkg/channel/consolidated/kafka/admin_test.go:86 +0x48
  knative.dev/eventing-kafka/pkg/channel/consolidated/kafka.TestAdminClient.func2()
      /home/abd4lla/Workspace/redhat/src/knative.dev/eventing-kafka/pkg/channel/consolidated/kafka/admin_test.go:63 +0x50

Goroutine 13 (running) created at:
  knative.dev/eventing-kafka/pkg/channel/consolidated/kafka.TestAdminClient()
      /home/abd4lla/Workspace/redhat/src/knative.dev/eventing-kafka/pkg/channel/consolidated/kafka/admin_test.go:62 +0x2c9
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:1127 +0x202
```

Reply:
ah wow, that's pretty bad